### ISSUE TITLE
Question on using donut_amd64 executor

### ISSUE THREAD

**Issue by kartikdurg** [mitre/caldera#1763]
Hello,

Even after including the donut functionality in my powershell snippet, I am not able to execute the shellcode on the victim machine.

I passed bytes of donut shellcode in the command field of the ability but, nothing seems to execute.
Am I doing it wrong? If so, it would be great if you can help me with an example. :)

Thank you!

**Comment by TheWover** [comment]
Hi!

Are you putting the shellcode directly into the ability? That functionality is intended for the `shellcode` executor rather than the `donut` executor. The Donut executor allows you to execute dynamically compiled C# code like in this ability: https://github.com/mitre/stockpile/blob/master/data/abilities/discovery/5a39d7ed-45c9-4a79-b581-e5fb99e24f65.yml

**Comment by TheWover** [comment]
If you are manually generating shellcode with Donut then you can put that into an ability that uses the shellcode executor like this one: https://github.com/mitre/stockpile/blob/master/data/abilities/privilege-escalation/0821b0b0-7902-4a7b-8052-80bda5a43684.yml

You can get output from Donut in that format by using the C#/Ruby output format available in Donut and then removing the variable declaration in the output file.

**Comment by kartikdurg** [comment]
Ahh cool. I thought the donut executor accepts bytes like the shellcode executor. 

But, Should I enable the "builder" plugin for the Donut executor to compile the C# code dynamically?

**Comment by TheWover** [comment]
@kartikdurg Yes, you have to enable the builder plugin to use the dynamic building.

**Comment by kartikdurg** [comment]
@TheWover Okay. Thank you! :)

**Comment by kartikdurg** [comment]
I enabled the builder plugin and executed the below ability:
[](https://github.com/mitre/stockpile/blob/master/data/abilities/discovery/5a39d7ed-45c9-4a79-b581-e5fb99e24f65.yml)

This "operation" never completes after executing the first two commands,i.e "Get-Process" & "tasklist". 

![caldera_donut_amd64](https://user-images.githubusercontent.com/20142153/85556302-577c0b80-b644-11ea-804f-ef41ca554046.JPG)

Output from "server.py":

`user@kali-caldera:~/caldera$ sudo python3 server.py --insecure --fresh
2020-06-24 04:06:35 - WARNING (server.py:97 <module>) --insecure flag set. Caldera will use the default.yml config file.
2020-06-24 04:06:35 - INFO  (server.py:104 <module>) Using main config from conf/default.yml
2020-06-24 04:06:35 - DEBUG (selector_events.py:59 __init__) Using selector: EpollSelector
2020-06-24 04:06:35 - DEBUG (learning_svc.py:19 __init__) Loaded 2 parsers
2020-06-24 04:06:35 - DEBUG (data_svc.py:62 restore_state) There are 0 jobs in the scheduler
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: http
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: udp
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: tcp
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: websocket
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: html
2020-06-24 04:06:35 - DEBUG (contact_svc.py:34 register) Registered contact: gist
2020-06-24 04:06:35 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: compass
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: training
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: response
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: access
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: stockpile
2020-06-24 04:06:36 - DEBUG (config.py:21 find_config_file) Trying paths: ['/root/.docker/config.json', '/root/.dockercfg']
2020-06-24 04:06:36 - DEBUG (config.py:25 find_config_file) Found file at path: /root/.docker/config.json
2020-06-24 04:06:36 - DEBUG (auth.py:182 load_config) Found 'auths' section
2020-06-24 04:06:36 - DEBUG (auth.py:142 parse_auth) Found entry (registry='https://index.docker.io/v1/', username='*********')
2020-06-24 04:06:36 - DEBUG (build_svc.py:61 _stage_build_dir) Build directory for csharp already constructed
2020-06-24 04:06:36 - DEBUG (connectionpool.py:428 _make_request) http://localhost:None "GET /v1.35/images/json?filter=mono%3Alatest&only_ids=0&all=0 HTTP/1.1" 200 251
2020-06-24 04:06:36 - DEBUG (connectionpool.py:428 _make_request) http://localhost:None "GET /v1.35/images/sha256:b074d513c5cdfadffc24239c00962926b7b965e64cbc06b7c8ad4d110d77f4cd/json HTTP/1.1" 200 None
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: builder
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: gameboard
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: sandcat
2020-06-24 04:06:36 - DEBUG (app_svc.py:104 load_plugins) Enabled plugin: manx
2020-06-24 04:06:36 - ERROR (data_svc.py:124 _load_abilities) Ability=286dffd3-153f-49f2-8ac3-0b2900f7db30 has wrong tactic
2020-06-24 04:06:37 - DEBUG (auth_svc.py:54 apply) Created authentication group: blue
2020-06-24 04:06:37 - DEBUG (auth_svc.py:54 apply) Created authentication group: red
2020-06-24 04:06:37 - INFO  (server.py:66 run_tasks) All systems ready.
`

Am I missing anything?

**Comment by TheWover** [comment]
Okay. In the GUI when you select the ability, does the donut executor show up on the ability's details?

**Comment by kartikdurg** [comment]
@TheWover  Yes. It does..

**Comment by TheWover** [comment]
Okay, for now I would recommend removing the Powershell command from the ability file. From what I can tell it looks like CALDERA doesn't have an option to select which executor to use. That would be a useful feature...

**Comment by khyberspache** [comment]
@kartikdurg @TheWover The way we wrote the agent (as far as I can remember), caldera will select a prioritzed option in the ability for each platform (psh -> windows, sh -> Darwin/linux).  This was initially configured in the `contact_svc` when a new agent object is created.  It might have been moved.  We did this to avoid conflicts and set us up to be able to (in the future) both manually and autonomously flip executors during an operation.
